### ISSUE TITLE
error: produced an unexpected new value for was present, but now absent.

### ISSUE THREAD

**Issue by chen23** [F5Networks/terraform-provider-bigip#305]
getting the following error when running bigip_as3
```
Error: Provider produced inconsistent result after apply

When applying changes to bigip_as3.common, provider
"registry.terraform.io/-/bigip" produced an unexpected new value for was
present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```
here's is the terraform code
```
provider "bigip" {
  address  = "https://${var.bigip1_ip}:443"
  username = "admin"
  password = "${file("password")}"
}

# deploy application using as3
resource "bigip_as3" "common" {
  as3_json    = "${file("common-as3.json")}"
}
```
and the AS3 declaration
```
{"class":"ADC","schemaVersion":"3.0.0","Common":{"class":"Tenant","Shared":{"class":"Application","template":"shared","SharedVIP":{"class":"Service_Address","virtualAddress":"10.1.10.10"},"DataCenterOne":{"class":"GSLB_Data_Center"},"OCPDNS":{"class":"GSLB_Server","dataCenter":{"use":"DataCenterOne"},"devices":[{"address":"10.1.20.240","addressTranslation":"10.1.20.240"}],"virtualServers":[{"address":"10.1.10.10","addressTranslation":"10.1.10.10","addressTranslationPort":443,"port":443},{"address":"10.1.10.100","addressTranslation":"10.1.10.100","addressTranslationPort":80,"port":80},{"address":"10.1.10.101","addressTranslation":"10.1.10.101","addressTranslationPort":80,"port":80},{"address":"10.1.10.102","addressTranslation":"10.1.10.102","addressTranslationPort":80,"port":80},{"address":"10.1.10.10","addressTranslation":"10.1.10.10","addressTranslationPort":80,"port":80}]},"constants":{"class":"Constants","fast":{"template":"openshift/0_shared","view":{"self_address":"10.1.20.240","virtual_address":"10.1.10.10","virtual_address1":"10.1.10.100","virtual_address2":"10.1.10.101","virtual_address3":"10.1.10.102"},"lastModified":"2020-07-01T10:24:07.108Z"}}}},"id":"fast-Common-Shared-bb690409-babf-4263-ac2e-19d4e44861ac","updateMode":"selective","controls":{"archiveTimestamp":"2020-07-01T10:24:13.526Z"}}
```
If i manually send the declaration it works w/out a problem
```
 curl -u admin:[password] -k https://[IP Address]/mgmt/shared/appsvcs/declare  -d @./common-as3.json -H expect:
{"results":[{"code":200,"message":"success","lineCount":21,"host":"localhost","tenant":"Common","runTime":1527},{"code":200,"message":"no change","host":"localhost","tenant":"Common","runTime":609}],"declaration":{"class":"ADC","schemaVersion":"3.0.0","Common":{"class":"Tenant","Shared":{"class":"Application","template":"shared","SharedVIP":{"class":"Service_Address","virtualAddress":"10.1.10.10"},"DataCenterOne":{"class":"GSLB_Data_Center"},"OCPDNS":{"class":"GSLB_Server","dataCenter":{"use":"DataCenterOne"},"devices":[{"address":"10.1.20.240","addressTranslation":"10.1.20.240"}],"virtualServers":[{"address":"10.1.10.10","addressTranslation":"10.1.10.10","addressTranslationPort":443,"port":443},{"address":"10.1.10.100","addressTranslation":"10.1.10.100","addressTranslationPort":80,"port":80},{"address":"10.1.10.101","addressTranslation":"10.1.10.101","addressTranslationPort":80,"port":80},{"address":"10.1.10.102","addressTranslation":"10.1.10.102","addressTranslationPort":80,"port":80},{"address":"10.1.10.10","addressTranslation":"10.1.10.10","addressTranslationPort":80,"port":80}]},"constants":{"class":"Constants","fast":{"template":"openshift/0_shared","view":{"self_address":"10.1.20.240","virtual_address":"10.1.10.10","virtual_address1":"10.1.10.100","virtual_address2":"10.1.10.101","virtual_address3":"10.1.10.102"},"lastModified":"2020-07-01T10:24:07.108Z"}}}},"id":"fast-Common-Shared-bb690409-babf-4263-ac2e-19d4e44861ac","updateMode":"selective","controls":{"archiveTimestamp":"2020-07-01T13:05:41.628Z"}}}
```

**Comment by RavinderReddyF5** [comment]
@chen23 Is it possible to share debug logs by setting environment variable `TF_LOG=DEBUG` ?

**Comment by chen23** [comment]
here's the output
```
2020-07-01T10:31:20.994-0400 [DEBUG] plugin: plugin started: path=/home/backup/projects/github/f5-k8s-demo/ocp4-aws-upi/deploy/.terraform/plugins/linux_amd64/terraform-provider-bigip_v1.2.1_x4 pid=23254
2020-07-01T10:31:20.994-0400 [DEBUG] plugin: waiting for RPC address: path=/home/backup/projects/github/f5-k8s-demo/ocp4-aws-upi/deploy/.terraform/plugins/lin
ux_amd64/terraform-provider-bigip_v1.2.1_x4
2020-07-01T10:31:21.003-0400 [INFO]  plugin.terraform-provider-bigip_v1.2.1_x4: configuring server automatic mTLS: timestamp=2020-07-01T10:31:21.002-0400     2020-07-01T10:31:21.035-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: plugin address: network=unix address=/tmp/plugin388177699 timestamp=2020-07-01
T10:31:21.035-0400
2020-07-01T10:31:21.035-0400 [DEBUG] plugin: using plugin: version=5
2020-07-01T10:31:21.112-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:21 [INFO] Initializing BigIP connection
2020/07/01 10:31:21 [WARN] Provider "registry.terraform.io/-/bigip" produced an invalid plan for bigip_as3.common, but we are tolerating it because it is usin
g the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
n            },\n            \"constants\": {\n                \"class\": \"Constants\",\n                \"fast\": {\n                    \"template\": \"ope
nshift/0_shared\",\n                    \"view\": {\n                        \"self_address\": \"10.1.20.240\",\n                        \"virtual_address\":
\"10.1.10.10\",\n                        \"virtual_address1\": \"10.1.10.100\",\n                        \"virtual_address2\": \"10.1.10.101\",\n
           \"virtual_address3\": \"10.1.10.102\"\n                    },\n                    \"lastModified\": \"2020-07-01T10:24:07.108Z\"\n
}\n            }\n        }\n    },\n    \"id\": \"fast-Common-Shared-bb690409-babf-4263-ac2e-19d4e44861ac\",\n    \"updateMode\": \"selective\",\n    \"contr
ols\": {\n        \"archiveTimestamp\": \"2020-07-01T10:24:13.526Z\"\n    }\n}\n")
bigip_as3.common: Creating...
2020/07/01 10:31:21 [DEBUG] bigip_as3.common: applying the planned Create change
2020-07-01T10:31:21.251-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:21 [INFO] Creating As3 config
2020-07-01T10:31:26.672-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:26 [DEBUG] AS3 Version:3.20.0
2020-07-01T10:31:26.672-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:26 [DEBUG] Terraform Version:Terraform/0.12.24
2020-07-01T10:31:26.672-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:26 [INFO] Creating as3 config in bigip:{"Common":{"Shared":{"
DataCenterOne":{"class":"GSLB_Data_Center"},"OCPDNS":{"class":"GSLB_Server","dataCenter":{"use":"DataCenterOne"},"devices":[{"address":"10.1.20.240","addressT
ranslation":"10.1.20.240"}],"virtualServers":[{"address":"10.1.10.10","addressTranslation":"10.1.10.10","addressTranslationPort":443,"port":443},{"address":"1
0.1.10.100","addressTranslation":"10.1.10.100","addressTranslationPort":80,"port":80},{"address":"10.1.10.101","addressTranslation":"10.1.10.101","addressTran
slationPort":80,"port":80},{"address":"10.1.10.102","addressTranslation":"10.1.10.102","addressTranslationPort":80,"port":80},{"address":"10.1.10.10","address
Translation":"10.1.10.10","addressTranslationPort":80,"port":80}]},"SharedVIP":{"class":"Service_Address","virtualAddress":"10.1.10.10"},"class":"Application"
,"constants":{"class":"Constants","fast":{"lastModified":"2020-07-01T10:24:07.108Z","template":"openshift/0_shared","view":{"self_address":"10.1.20.240","virt
ual_address":"10.1.10.10","virtual_address1":"10.1.10.100","virtual_address2":"10.1.10.101","virtual_address3":"10.1.10.102"}}},"template":"shared"},"class":"
Tenant","controls":{"class":"Controls","userAgent":"Terraform/0.12.24"}},"class":"ADC","controls":{"archiveTimestamp":"2020-07-01T10:24:13.526Z","controls":{"
class":"Controls","userAgent":"Terraform/0.12.24"}},"id":"fast-Common-Shared-bb690409-babf-4263-ac2e-19d4e44861ac","schemaVersion":"3.0.0","updateMode":"selec
tive"}
2020-07-01T10:31:26.717-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:26 [DEBUG]Code = 0,ID = 6f6debc1-1bd6-4170-9c43-a2b34dc804e9
2020-07-01T10:31:29.733-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:29 [DEBUG]Sucessfully Created Application with ID  = 6f6debc1
-1bd6-4170-9c43-a2b34dc804e9                                                                                                                                  2020-07-01T10:31:29.733-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:29 [INFO] Reading As3 config
2020-07-01T10:31:30.067-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:30 [ERROR] Unable to retrieve json
2020-07-01T10:31:30.067-0400 [DEBUG] plugin.terraform-provider-bigip_v1.2.1_x4: 2020/07/01 10:31:30 [ERROR] unexpected end of JSON input
2020/07/01 10:31:30 [DEBUG] bigip_as3.common: apply errored, but we're indicating that via the Error pointer rather than returning it: Provider produced incon
sistent result after apply: When applying changes to bigip_as3.common, provider "registry.terraform.io/-/bigip" produced an unexpected new value for was prese
nt, but now absent.
```

**Comment by RavinderReddyF5** [comment]
@chen23 it looks like because of empty tenant_list we failed to retrieve as3 json from bigip in as3 resource read call. Created TER-400 to track issue.

**Comment by chen23** [comment]
If I changed the input to add the top-level "class":"AS3" it works as expected.  This could be just a documentation issue to specify that requirement and/or handle the error more gracefully.

new input that works:
```
{"class":"AS3","action":"deploy","persist":true,"declaration":{"class":"ADC","schemaVersion":"3.0.0","Common":{"class":"Tenant","Shared":{"class":"Application","template":"shared","SharedVIP":{"class":"Service_Address","virtualAddress":"10.1.10.10"},"DataCenterOne":{"class":"GSLB_Data_Center"},"OCPDNS":{"class":"GSLB_Server","dataCenter":{"use":"DataCenterOne"},"devices":[{"address":"10.1.20.240","addressTranslation":"10.1.20.240"}],"virtualServers":[{"address":"10.1.10.10","addressTranslation":"10.1.10.10","addressTranslationPort":443,"port":443},{"address":"10.1.10.100","addressTranslation":"10.1.10.100","addressTranslationPort":80,"port":80},{"address":"10.1.10.101","addressTranslation":"10.1.10.101","addressTranslationPort":80,"port":80},{"address":"10.1.10.102","addressTranslation":"10.1.10.102","addressTranslationPort":80,"port":80},{"address":"10.1.10.10","addressTranslation":"10.1.10.10","addressTranslationPort":80,"port":80}]},"constants":{"class":"Constants","fast":{"template":"openshift/0_shared","view":{"self_address":"10.1.20.240","virtual_address":"10.1.10.10","virtual_address1":"10.1.10.100","virtual_address2":"10.1.10.101","virtual_address3":"10.1.10.102"},"lastModified":"2020-07-01T10:24:07.108Z"}}}},"id":"fast-Common-Shared-bb690409-babf-4263-ac2e-19d4e44861ac","updateMode":"selective","controls":{"archiveTimestamp":"2020-07-01T10:24:13.526Z"}}}
```

**Comment by RavinderReddyF5** [comment]
@chen23 Thanks for your validation with addition of AS3 class, we will add validation on input json to provide it from top level AS3 class.

**Comment by focrensh** [comment]
Tracking with TER-400

**Comment by papineni87** [comment]
Issue is fixed in 1.3 release, closing the issue
